clarity/src/vm/analysis/type_checker/natives/sequences.rs

```diff
@@ -423,3 +423,28 @@ pub fn check_special_slice(
 
     Ok(seq)
 }
+
+/// This function type checks the Clarity2 function `replace-at`.
+pub fn check_special_replace_at(
+    checker: &mut TypeChecker,
+    args: &[SymbolicExpression],
+    context: &TypingContext,
+) -> TypeResult {
+    check_argument_count(3, args)?;
+
+    runtime_cost(ClarityCostFunction::AnalysisIterableFunc, checker, 0)?;
+    // Check sequence
+    let seq_type = checker.type_check(&args[0], context)?;
+    let unit_seq = match &seq_type {
+        TypeSignature::SequenceType(seq) => seq.unit_type(),
+        _ => return Err(CheckErrors::ExpectedSequence(seq_type).into()),
+    };
+
+    // Check index argument
+    checker.type_check_expects(&args[1], context, &TypeSignature::UIntType)?;
+    // Check element argument
+    let elem_type = checker.type_check(&args[2], context)?;
+    TypeSignature::least_supertype(&unit_seq, &elem_type)?;
+
+    Ok(seq_type)
```

Comment:
I think you want to use the `least_supertype` result to create the new sequence type in the case of a list. In the case of Buffers and Strings, I think you don't want to use `least_supertype` at all, because that will "accept" a `(ascii-string 2)` as an element. This leads to the length issues only being caught at runtime (as you show in your tests).  That's not usually how other Clarity functions work (or, if they do, like `as-max-len?`, they return an option or result type for the check).

```suggestion
    let input_type = checker.type_check(&args[0], context)?;
    let seq_type = match input_type {
        TypeSignature::SequenceType(seq) => seq,
        _ => return Err(CheckErrors::ExpectedSequence(input_type).into()),
    };
    let unit_seq = seq_type.unit_type();

    // Check index argument
    checker.type_check_expects(&args[1], context, &TypeSignature::UIntType)?;
    // Check element argument
    match &seq_type {
        ListType(list_type) => {
            let replace_type = checker.type_check(&args[2], context)?;
            let new_unit_type = TypeSignature::least_supertype(&unit_seq, &replace_type)?;
            let out_type = TypeSignature::list_of(new_unit_type, list_type.get_max_len())?;
            Ok(out_type)
        }
        BufferType(_) | StringType(_) => {
            checker.type_check_expects(&args[2], context, &unit_seq)?;
            Ok(TypeSignature::SequenceType(seq_type))
        }
    }
```

That way, length mismatches will be caught for `buff` and `string` types, and lists will get the proper supertype. For example:

```
(replace-at (list 0x00 0x0000 0x0000) 1 0x000000)
```

Will have type `(list (buff 3) 3)`. Right now, I think that would incorrectly return `(list (buff 2) 3)`.

I think this would also require an update to the runtime implementation. Alternatively, if you think that `replace-at` should always preserve the input list's type, then the type checker can do away with the call to `least_supertype`, and instead you `type_check_expects(.., .., &unit_seq)` in all cases.

Reply:
I think I was thinking the latter approach makes the most sense (preserving the input list's type).